### PR TITLE
Add `FromBytes::read_from_io` and `IntoBytes::write_to_io`

### PR DESCRIPTION
**Pull request by jswrenn**
Makes progress on #158.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -4525,6 +4529,24 @@ pub unsafe trait FromBytes: FromZeros {
         }
     }
 
+    /// TODO
```
Comment (joshlf) [inline src/lib.rs:4532]: TODO: Doc comment

File: src/lib.rs [hunk 2]
```diff
@@ -5188,6 +5210,16 @@ pub unsafe trait IntoBytes {
         Ok(())
     }
 
+    /// TODO
```
Comment (joshlf) [inline src/lib.rs:5213]: TODO: Doc comment

File: src/lib.rs
Comment (joshlf) [inline src/lib.rs:1]: TODO: Tests

**Review by joshlf — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 3]
```diff
@@ -4525,6 +4528,48 @@ pub unsafe trait FromBytes: FromZeros {
         }
     }
 
+    /// Reads a copy of `self` from the IO byte source `src`.
```
Comment (joshlf) [inline src/lib.rs:4531]:
```suggestion
    /// Reads a copy of `self` from an `io::Read`.
```

File: src/lib.rs [hunk 4]
```diff
@@ -4525,6 +4528,48 @@ pub unsafe trait FromBytes: FromZeros {
...
+    ///
+    /// This is useful for interfacing with operating system byte sinks (files,
+    /// sockets, etc.).
```
Comment (joshlf) [inline src/lib.rs:4534]: IMO this is more specific than we need - non-OS resources can implement `io::Read` too. I think we can just remove this sentence.

File: src/lib.rs [hunk 5]
```diff
@@ -4525,6 +4528,48 @@ pub unsafe trait FromBytes: FromZeros {
...
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
```
Comment (joshlf) [inline src/lib.rs:4538]: If you want this to run, you could use a trick like we do here: https://github.com/google/zerocopy/blob/ace5bb0de76e150d2284984677892a5aaa34a36e/src/macros.rs#L711

File: src/lib.rs [hunk 6]
```diff
@@ -5188,6 +5233,55 @@ pub unsafe trait IntoBytes {
...
     }
 
+    /// Writes a copy of `self` to the IO byte sink `dst`.
+    ///
+    /// This is a shorthand for `dst.write_all(self.as_bytes())`, and is useful
+    /// for interfacing with operating system byte sinks (files, sockets, etc.).
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
```
Comment (joshlf) [inline src/lib.rs:5243]: If you want this to run, you could use a trick like we do here: https://github.com/google/zerocopy/blob/ace5bb0de76e150d2284984677892a5aaa34a36e/src/macros.rs#L711

File: src/lib.rs [hunk 7]
```diff
@@ -5188,6 +5233,55 @@ pub unsafe trait IntoBytes {
         Ok(())
     }
 
+    /// Writes a copy of `self` to the IO byte sink `dst`.
```
Comment (joshlf) [inline src/lib.rs:5236]:
```suggestion
    /// Writes a copy of `self` to an `io::Write`.
```

File: src/lib.rs [hunk 8]
```diff
@@ -5188,6 +5233,55 @@ pub unsafe trait IntoBytes {
...
+    ///
+    /// This is a shorthand for `dst.write_all(self.as_bytes())`, and is useful
+    /// for interfacing with operating system byte sinks (files, sockets, etc.).
```
Comment (joshlf) [inline src/lib.rs:5239]: Same here - IMO we should remove the mention of OS byte sinks.

File: src/lib.rs [hunk 9]
```diff
@@ -5188,6 +5233,55 @@ pub unsafe trait IntoBytes {
...
+    /// ```
+    ///
+    /// If the write fails, `write_to_io` returns `Err` and a partial write may
+    /// have occured; e.g.:
+    ///
+    /// ```
+    /// # use zerocopy::IntoBytes;
+    ///
+    /// let src = u128::MAX;
+    /// let mut dst = [0u8, 0];
```
Comment (joshlf) [inline src/lib.rs:5268]:
```suggestion
    /// let mut dst = [0u8; 2];
```

I misread this as `[0u8; 0]` and was confused by what was being tested. IMO this is less subject to being misread.

**Review by joshlf — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 4 above)
Comment (jswrenn) [reply src/lib.rs:4534]: This is here solely for the benefit of SEO/Cmd+F. I don't say that these are the *only* uses for this method, but while non-OS resources *can* implement `io::Read`, you're almost *always* better off using our other methods in those cases.

It's not a load bearing sentence, but I think it's a helpful one.

Comment (joshlf) [reply src/lib.rs:4534]: Ack.

File: src/lib.rs (hunk 6 above)
Comment (jswrenn) [reply src/lib.rs:5243]: I don't think there's much sense in actually testing the IO. It'll make our tests more expensive to run, and will cause them to fail on platforms where file IO is stubbed out (e.g., wasm).

Comment (joshlf) [reply src/lib.rs:5240]: Ack

**Review by joshlf — APPROVED** [review]

**Comment by joshlf** [comment]
Backporting in https://github.com/google/zerocopy/pull/2046

**Pull request merged**